File: src/main/java/cz/encircled/eplayer/service/JsonCacheService.java
package cz.encircled.eplayer.service;

import com.google.gson.JsonSyntaxException;
import cz.encircled.eplayer.core.ApplicationCore;
import cz.encircled.eplayer.model.MediaType;
import cz.encircled.eplayer.util.IOUtil;
import cz.encircled.eplayer.util.SettingsProvider;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.*;
import java.util.function.Consumer;

/**
 * @author Encircled on 9.6.2014.
 */
public class JsonCacheService implements CacheService {

    public static final String QUICK_NAVI_PATH = ApplicationCore.APP_DOCUMENTS_ROOT + "quicknavi2.json";
    private static final Logger log = LogManager.getLogger();
    private Map<String, MediaType> cache = new HashMap<>();

    public void init() {
        long start = System.currentTimeMillis();
        log.trace("JsonCacheService init start");
        if (IOUtil.createIfMissing(QUICK_NAVI_PATH, false, true)) {
            log.debug("QuickNavi file was created");
        }
        try {
            cache = IOUtil.getPlayableJson(QUICK_NAVI_PATH);
        } catch (IOException e) {
            log.error("Failed to read cache data from {} with default type token. Message: {}",
                    SettingsProvider.get(QUICK_NAVI_PATH), e.getMessage());
            // TODO
//            guiUtil.showMessage(msgQnFileIoFail.ln(), errorTitle.ln());
        } catch (JsonSyntaxException e) {
            log.error("JSON syntax error. Message: {}", e.getMessage());
            // TODO
//            guiUtil.showMessage(msgQnFileCorrupted.ln(), errorTitle.ln());
        }
        if (cache == null) {
            cache = new HashMap<>();
        }

        log.trace("JsonCacheService init complete in {} ms", System.currentTimeMillis() - start);
    }

    @Override
    public void forEach(@NotNull Consumer<MediaType> action) {
        getCache().forEach(action);
    }

    @NotNull
    @Override
    public MediaType createIfAbsent(@NotNull String path) {
        return createIfAbsent(path, null);
    }

    @NotNull
    @Override
    public MediaType createIfAbsent(@NotNull String path, @Nullable String title) {
        return cache.computeIfAbsent(path, p -> new MediaType(path, title));
    }

    @NotNull
    @Override
    public MediaType addIfAbsent(@NotNull MediaType mediaType) {
        log.debug("Add if absent {}", mediaType.toString());
        return cache.putIfAbsent(mediaType.getPath(), mediaType);
    }

    @Override
    @Nullable
    public MediaType getEntry(@NotNull String id) {
        return cache.get(id);
    }

    @Override
    @Nullable
    public MediaType updateEntry(@NotNull String id, long time) {
        MediaType p = cache.get(id);
        log.debug("Update cache entry {}, time {}", p, time);
        p.setTime(time);
        p.setWatchDate(new Date().getTime());
        return p;
    }

    @Override
    @Nullable
    public MediaType deleteEntry(@NotNull String id) {
        return cache.remove(id);
    }

    @NotNull
    @Override
    public List<MediaType> getCache() {
        return new ArrayList<>(cache.values());
    }

    @Nullable
    @Override
    public MediaType getLastByWatchDate() {
        MediaType p = getCache()
                .stream()
                .max(Comparator.comparingLong(MediaType::getWatchDate)).get();
        log.debug("Last played: {}", p);
        return p;
    }

    /**
     * Save playable map to file in JSON format
     */
    @Override
    public synchronized void save() {
        log.debug("Save json cache");
        try {
            IOUtil.storeJson(cache, QUICK_NAVI_PATH);
            log.debug("Json successfully saved");
        } catch (IOException e) {
            log.error("Failed to save playable to {}, msg {}", QUICK_NAVI_PATH, e);
        }
    }

}

File: src/main/java/cz/encircled/eplayer/service/JsonCacheService.kt
package cz.encircled.eplayer.service

import cz.encircled.eplayer.core.ApplicationCore
import cz.encircled.eplayer.model.MediaFile
import cz.encircled.eplayer.model.MediaSeries
import cz.encircled.eplayer.model.PlayableMedia
import cz.encircled.eplayer.model.SingleMedia
import cz.encircled.eplayer.service.event.Event
import cz.encircled.eplayer.util.Localization
import cz.encircled.eplayer.util.TimeMeasure.measure
import org.apache.logging.log4j.LogManager
import java.io.File
import java.io.IOException

/**
 * @author Encircled on 9.6.2014.
 */
class JsonCacheService(val core: ApplicationCore) : CacheService {

    private lateinit var cache: MutableMap<String, PlayableMedia>

    init {
        measure("JsonCacheService init") {
            try {
                cache = core.ioUtil.getPlayableJson()
                    .map {
                        core.metaInfoService.fetchMetadataAsync(it)
                    }
                    .associateBy { it.path }
                    .toMutableMap()
            } catch (e: Exception) {
                cache = HashMap()
                log.error("Failed to read cache data", e)
                core.appView.showUserMessage(Localization.msgQnFileIoFail.ln())
            }
        }

        measure("JsonCacheService init listeners") {
            addEventListeners()
        }
    }

    private fun addEventListeners() {
        Event.mediaTimeChange.listen {
            cache.getValue(it.playableMedia.path).time.set(it.characteristic)
        }
        Event.mediaDurationChange.listen {
            cache.getValue(it.playableMedia.path).duration.set(it.characteristic)
        }

        Event.subtitleChanged.listen {
            if (it.changedByUser) cache.getValue(it.playableMedia.path).preferredSubtitle = it.characteristic
        }

        Event.audioTrackChanged.listen {
            if (it.changedByUser) cache.getValue(it.playableMedia.path).preferredAudio = it.characteristic
        }

        Event.metadataAcquired.listen {
            // TODO for series
            val media = cache[it.playableMedia.path]
            if (media is SingleMedia) {
                media.metaCreationDate = it.characteristic.getOrDefault("creation_time", "")
            }
        }
    }

    override fun getOrNull(path: String): PlayableMedia? {
        val mediaFile = MediaFile(path)

        return if (core.seriesFinder.isSeries(mediaFile.name)) {
            cache[core.seriesFinder.seriesName(mediaFile)]
        } else {
            cache[path]
        }
    }

    override fun createIfAbsent(path: String): PlayableMedia {
        val mediaFile = MediaFile(path)

        return if (core.seriesFinder.isSeries(mediaFile.name)) {
            getOrCreateSeries(mediaFile, path)
        } else {
            cache.getOrPut(path) {
                createNewMedia(path)
            }
        }
    }

    override fun createIfAbsent(media: PlayableMedia): PlayableMedia = cache.getOrPut(media.path) { media }

    override fun deleteEntry(media: PlayableMedia): PlayableMedia = cache.remove(media.path)!!

    override fun getCachedMedia(): List<PlayableMedia> = ArrayList(cache.values)

    override fun getPlayedMedia(): List<PlayableMedia> = getCachedMedia().filter { it.isPlayed() }

    override fun lastByWatchDate(): PlayableMedia? {
        val p = getCachedMedia().maxWithOrNull(Comparator.comparingLong(PlayableMedia::watchDate))
        log.debug("Last played: {}", p)
        return p
    }

    /**
     * Save playable map to file in JSON format
     */
    @Synchronized
    override fun save() {
        log.debug("Save json cache")
        try {
            core.ioUtil.savePlayable(cache)
            log.debug("Json successfully saved")
        } catch (e: IOException) {
            log.error("Failed to save playable, msg {}", e)
        }
    }

    private fun getOrCreateSeries(mediaFile: MediaFile, path: String): PlayableMedia {
        val seriesName = core.seriesFinder.seriesName(mediaFile)
        val seriesPath = File(mediaFile.path).parent + File.separator + seriesName

        if (cache.containsKey(seriesPath)) {
            val mediaSeries = cache.getValue(seriesPath) as MediaSeries
            if (!mediaSeries.series.map { it.path }.contains(path)) {
                mediaSeries.series.add(createNewMedia(path))
            }
        } else {
            val allSeries = core.seriesFinder.findSeriesForName(File(mediaFile.path).parent, seriesName)
                .map { createNewMedia(it) }


            cache[seriesPath] = MediaSeries(seriesName, seriesPath, ArrayList(allSeries))
        }

        return cache.getValue(seriesPath)
    }

    private fun createNewMedia(path: String): SingleMedia {
        log.debug("createNewMedia $path")
        return core.metaInfoService.fetchMetadataAsync(SingleMedia(path))
    }

    companion object {
        private val log = LogManager.getLogger()
    }
}